// @composio/client/resources/org/project/project.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Project = void 0;
const tslib_1 = require("../../../internal/tslib.js");
const resource_1 = require("../../../core/resource.js");
const APIKeysAPI = tslib_1.__importStar(require("./api-keys.js"));
const api_keys_1 = require("./api-keys.js");
const TriggerAPI = tslib_1.__importStar(require("./trigger.js"));
const trigger_1 = require("./trigger.js");
const WebhookAPI = tslib_1.__importStar(require("./webhook.js"));
const webhook_1 = require("./webhook.js");
const path_1 = require("../../../internal/utils/path.js");
class Project extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.apiKeys = new APIKeysAPI.APIKeys(this._client);
        this.webhook = new WebhookAPI.Webhook(this._client);
        this.trigger = new TriggerAPI.Trigger(this._client);
    }
    /**
     * Creates a new project within the authenticated user's organization using the
     * specified name. Projects are isolated environments within your organization,
     * each with their own API keys, webhook configurations, and resources. Use this
     * endpoint to create additional projects for different environments (e.g.,
     * development, staging, production) or for separate applications.
     *
     * @example
     * ```ts
     * const project = await client.org.project.create({
     *   name: 'my-production-api',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/v3/org/project/new', { body, ...options });
    }
    /**
     * Retrieves detailed information about a specific project using its unique
     * identifier. This endpoint provides complete project configuration including
     * webhook URLs, creation and update timestamps, and webhook secrets. Use this
     * endpoint to inspect project settings or verify project configuration.
     *
     * @example
     * ```ts
     * const project = await client.org.project.retrieve(
     *   'proj_abc123xyz456',
     * );
     * ```
     */
    retrieve(projectID, options) {
        return this._client.get((0, path_1.path) `/api/v3/org/project/${projectID}`, options);
    }
    /**
     * Retrieves all projects belonging to the authenticated user's organization.
     * Projects are returned in descending order of creation date (newest first). This
     * endpoint is useful for displaying project selection in dashboards or for
     * integrations that need to list all available projects.
     *
     * @example
     * ```ts
     * const projects = await client.org.project.list();
     * ```
     */
    list(query = {}, options) {
        return this._client.get('/api/v3/org/project/list', { query, ...options });
    }
    /**
     * Soft-deletes a project within the organization by its unique identifier. When a
     * project is deleted, it is marked as deleted but not immediately removed from the
     * database. This operation affects all resources associated with the project
     * including API keys, webhook configurations, and connected services. This action
     * cannot be undone through the API.
     *
     * @example
     * ```ts
     * const project = await client.org.project.delete(
     *   'proj_abc123xyz456',
     * );
     * ```
     */
    delete(projectID, options) {
        return this._client.delete((0, path_1.path) `/api/v3/org/project/delete/${projectID}`, options);
    }
}
exports.Project = Project;
Project.APIKeys = api_keys_1.APIKeys;
Project.Webhook = webhook_1.Webhook;
Project.Trigger = trigger_1.Trigger;
